pal/src/misc/random.cpp

```diff
@@ -17,35 +17,54 @@ Module Name:
 
 --*/
 
-#include "pal/palinternal.h"
-#include <sodium.h>
+#ifdef _CRT_RAND_S
 
-/*++
-Function :
-RANDOMInitialize
+#include "pal/palinternal.h"
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <mutex>
 
-Initialize libsodium
 
-(no parameters)
+#define RANDOM_CACHE_SIZE 8
+constexpr int rand_byte_len = sizeof(unsigned int) * RANDOM_CACHE_SIZE;
+static union {
+    char rstack[rand_byte_len];
+    unsigned int result[RANDOM_CACHE_SIZE];
+} random_cache;
+static unsigned cache_index(RANDOM_CACHE_SIZE);
+static std::mutex random_generator_mutex;
 
-Return value :
-TRUE  if libsodium initialization succeeded
-FALSE otherwise
---*/
-BOOL RANDOMInitialize(void)
+static bool GetRandom(unsigned int *result) noexcept
 {
-    if (sodium_init() == -1)
-    {
-        return FALSE;
+    std::lock_guard<std::mutex> guard(random_generator_mutex);
+    if (cache_index < RANDOM_CACHE_SIZE) {
+        *result = random_cache.result[cache_index++];
+        return true;
     }
-    
-    return TRUE;
-}
 
+    const int rdev = open("/dev/urandom", O_RDONLY);
+    unsigned len = 0;
 
-// Implemented in stdlib in the universal CRT
-errno_t __cdecl rand_s(unsigned int* randomValue)
+    do
+    {
+        int added = read(rdev, random_cache.rstack + len, rand_byte_len - len);
+        if (added < 0) {
+            close(rdev);
+            return false;
+        }
+        len += added;
+    } while (len < rand_byte_len);
+
+    close(rdev);
+    *result = random_cache.result[0];
+    cache_index = 1;
+    return len == rand_byte_len;
+}
+
+errno_t __cdecl rand_s(unsigned int* randomValue) noexcept
```

Comment:
We should initialize randomValue to 0 as per the docs for rand_s at https://msdn.microsoft.com/en-us/library/sxtz2fa8.aspx